src/ca/mcgill/ecse211/lab4/Lab4.java

```diff
@@ -0,0 +1,109 @@
+import lejos.hardware.sensor.SensorModes;
+import lejos.robotics.SampleProvider;
+import ca.mcgill.ecse211.odometer.Odometer;
+import ca.mcgill.ecse211.odometer.OdometerExceptions;
+
+
+
+public class Lab4 {
+
+  // Motor Objects, and Robot related parameters
+  private static final EV3LargeRegulatedMotor leftMotor =
+      new EV3LargeRegulatedMotor(LocalEV3.get().getPort("A"));
+  private static final EV3LargeRegulatedMotor rightMotor =
+      new EV3LargeRegulatedMotor(LocalEV3.get().getPort("B"));
+  private static final TextLCD lcd = LocalEV3.get().getTextLCD();
+  private static final Port usPort = LocalEV3.get().getPort("S1");
+
+  // Robot related parameters
+  public static final double WHEEL_RAD = 2.2;// the radius of the wheel
+  public static final double TRACK = 10.6;// the axel track of the robot
+
+  public static void main(String[] args) throws OdometerExceptions {
+
+    int buttonChoice;
+
+    // Odometer related objects
+    Odometer odometer = Odometer.getOdometer(leftMotor, rightMotor, TRACK, WHEEL_RAD);
+    // initialize the display
+    Display odometryDisplay = new Display(lcd);
+
+    @SuppressWarnings("resource") // Because we don't bother to close this resource
+    // Instance ultrasonicsensor
+    SensorModes ultrasonicSensor = new EV3UltrasonicSensor(usPort);
+    // usDistance fetch samples from this instance
+    SampleProvider usDistance = ultrasonicSensor.getMode("Distance");
+
+    UltrasonicLocalizer usLocalizer =
+        new UltrasonicLocalizer(odometer, leftMotor, rightMotor, usDistance);
+    LightLocalizer lsLocalizer = new LightLocalizer(odometer, leftMotor, rightMotor);
+
+
+    do {
+      // clear display
+      lcd.clear();
+
+      // ask the user whether the motors should use rising or falling edge
+      lcd.drawString("< Left | Right >", 0, 0);
+      lcd.drawString("       |        ", 0, 1);
+      lcd.drawString("Rising |Falling ", 0, 2);
+      lcd.drawString(" Edge  |  Edge  ", 0, 3);
+      lcd.drawString("       |        ", 0, 4);
+
+      buttonChoice = Button.waitForAnyPress();
+    } while (buttonChoice != Button.ID_LEFT && buttonChoice != Button.ID_RIGHT);
+
+    if (buttonChoice == Button.ID_LEFT) {// when the user choose rising edge
+
+      Thread odoThread = new Thread(odometer);
+      odoThread.start();
+      Thread odoDisplayThread = new Thread(odometryDisplay);
+      odoDisplayThread.start();
+
+      Thread lightThread = new Thread(lsLocalizer);
+      usLocalizer.risingEdge();// use rising edge
+      buttonChoice = Button.waitForAnyPress();// start the light localizer after pressing the button
+      lightThread.start();
+
+    } else {
+
+      Thread odoThread = new Thread(odometer);
+      odoThread.start();
+      Thread odoDisplayThread = new Thread(odometryDisplay);
+      odoDisplayThread.start();
+
+      usLocalizer.fallingEdge();// use falling edge
+      buttonChoice = Button.waitForAnyPress();
+      lsLocalizer.run();
+
+    }

```

Comment:
Repetitive code. Rewrite it like this:

```java
} while (buttonChoice != Button.ID_LEFT && buttonChoice != Button.ID_RIGHT);

new Thread(odometer).start();
new Thread(odometryDisplay).start();

if (buttonChoice == Button.ID_LEFT) { // when the user chooses rising edge
  UltrasonicLocalizer.risingEdge(); 
} else {
  UltrasonicLocalizer.fallingEdge();
}

buttonChoice = Button.waitForAnyPress(); // start the light localizer after pressing the button
LightLocalizer.localize();
```

You were inconsistent with your light localizer. In your if-clause, you ran it in a thread, but you simply called `run()` in the else. See below to understand why I changed to `lsLocalizer.run();` to `LightLocalizer.`_`localize`_`();`, a static call.